### ISSUE TITLE
Update to faster implementation

### ISSUE THREAD

**Issue by MichelleKendall** [thibautjombart/treescape#20]
I'll work on this on a separate fork tomorrow

**Comment by MichelleKendall** [comment]
On the newfeatures branch I have updated the "tree.vec" function to use Jacob's latest implementation, which relies on the function "CPP_update_combinations". However, I had trouble getting the C++ related bits working - solved some issues, but it is now failing the check with the message:

Error in .Primitive(".Call")(<pointer: (nil)>, length_root_distances,  : 
  NULL value passed as symbol address
Calls: exploratree ... lapply -> FUN -> CPP_update_combinations -> <Anonymous> 

Any ideas? Thanks!

**Comment by MichelleKendall** [comment]
Update: after doing more reading I think I was wrong to just use

#' @useDynlib 

so have changed it to

#' @useDynLib exploratree

However, I now get a different error (possibly specific to my machine?):

checking whether package 'exploratree' can be installed ...Warning: running command '"C:/PROGRA~1/R/R-31~1.3/bin/x64/Rcmd.exe" INSTALL -l "C:/Users/Michelle/AppData/Local/Temp/RtmpWeFI5q/exploratree.Rcheck" --no-html "C:\Users\Michelle\AppData\Local\Temp\RTMPWE~1\EXPLOR~1.RCH\00_PKG~1\EXPLOR~1"' had status 1
 [20s] ERROR
Installation failed.
See 'C:/Users/Michelle/AppData/Local/Temp/RtmpWeFI5q/exploratree.Rcheck/00install.out' for details.

The contents of that file are:
* installing *source* package 'exploratree' ...
** R
** data
*** moving datasets to lazyload DB
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
Error in library.dynam(lib, package, package.lib) : 
  DLL 'exploratree' not found: maybe not installed for this architecture?
Error: loading failed
Execution halted
*** arch - x64
Error in library.dynam(lib, package, package.lib) : 
  DLL 'exploratree' not found: maybe not installed for this architecture?
Error: loading failed
Execution halted
ERROR: loading failed for 'i386', 'x64'
* removing 'C:/Users/Michelle/AppData/Local/Temp/RtmpWeFI5q/exploratree.Rcheck/exploratree'

**Comment by MichelleKendall** [comment]
Sorry, I haven't managed to resolve this today. I think it's my use of "@useDynlib" - I get errors without it but I think I'm using it wrong. I will come back to it on Monday.

**Comment by thibautjombart** [comment]
I'm new to Rcpp, but my understanding is that using `cppFunction` compiles and links things on the fly. This means there is no shared objects to load when attaching the package; hence the error with useDynlib - there's nothing to load.

**Comment by thibautjombart** [comment]
The check now fails with:
```r
* checking examples ... ERROR
Running examples in ‘treescape-Ex.R’ failed
The error most likely occurred in:

> base::assign(".ptime", proc.time(), pos = "CheckExEnv")
> ### Name: tree.vec
> ### Title: Tree vector function
> ### Aliases: tree.vec
> 
> ### ** Examples
> 
> ## generate a random tree
> tree <- rtree(6)
> ## topological vector of mrca distances from root:
> tree.vec(tree)
Error in .Primitive(".Call")(<pointer: (nil)>, length_root_distances,  : 
  NULL value passed as symbol address
Calls: tree.vec ... lapply -> FUN -> CPP_update_combinations -> <Anonymous>
Execution halted
* checking PDF version of manual ... OK
* DONE

Status: 1 ERROR

```

**Comment by thibautjombart** [comment]
This is not a package error, but a coding error I think. I get:
```r
> traceback()
6: .Primitive(".Call")(<pointer: (nil)>, length_root_distances, 
       topological_root_distances, left_partition, right_partition, 
       index_offsets, distance_to_root, edges_to_root)
5: CPP_update_combinations(length_root_distances, topological_root_distances, 
       node$partitions[[group_a]], node$partitions[[group_b]], index_offsets, 
       node$root_distance, node$edges_to_root)
4: FUN(X[[i]], ...)
3: lapply(X = X, FUN = FUN, ...)
2: sapply(annotated_nodes, function(node) {
       if (length(node$partitions) > 1 && node$root_distance > 0) {
           num_groups <- length(node$partitions)
           for (group_a in 1:(num_groups - 1)) {
               for (group_b in (group_a + 1):num_groups) {
                   print(list(length_root_distances, topological_root_distances, 
                     node$partitions[[group_a]], node$partitions[[group_b]], 
                     index_offsets, node$root_distance, node$edges_to_root))
                   CPP_update_combinations(length_root_distances, 
                     topological_root_distances, node$partitions[[group_a]], 
                     node$partitions[[group_b]], index_offsets, 
                     node$root_distance, node$edges_to_root)
               }
           }
       }
   })
1: tree.vec(tre)
```

The arguments passed to the cpp procedure were:
```r
[[1]]
 [1] 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000
 [8] 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000 0.0000000
[15] 0.0000000 0.6386874 0.7187711 0.9550851 0.2137558 0.6171494 0.2414687

[[2]]
 [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1

[[3]]
[1] 2

[[4]]
[1] 1 4

[[5]]
[1]  0  5  9 12 14 15

[[6]]
[1] 0.9318219

[[7]]
[1] 1
```

**Comment by thibautjombart** [comment]
I think we may want to leave this for the 1.0.1 release on CRAN.

**Comment by MichelleKendall** [comment]
It's strange because it works locally so I don't understand how it could be a coding error. I'm going to create my own fork to have a go at fixing it. It would be really good to include this version in the first release because it's so much more usable and we wouldn't want anyone to dismiss using the package because they think they have too many trees / their trees are too big for it to be useful.

**Comment by thibautjombart** [comment]
So I tried locally, i.e. not within the example of the package, and got the
same error. Can you post a chunk of reproducible code that works for you?

**Comment by MichelleKendall** [comment]
Sure. This works for me:

tree.vec <- function(tree, lambda=0, return_lambda_function=F) {
  
  if(lambda<0 || lambda>1) stop("Pick lambda in [0,1]")
  
  num_leaves <- length(tree$tip.label)
  num_edges <- nrow(tree$edge)
  
  # We work with ordered labels, using this vector to transform indices.
  tip_order <- match(1:num_leaves, order(tree$tip.label))
  
  # Ordering the edges by first column places the root at the bottom.
  # Descendants will be placed always before parents.
  edge_order <- order(tree$edge[,1], decreasing=T)
  edges <- tree$edge[edge_order,]
  edge_lengths <- tree$edge.length[edge_order]
  
  # We annotated the nodes of the tree in this list. In two passes we are going to
  # compute the partition each node induces in the tips (bottom-up pass) and the distance 
  # (in branch length and number of branches) from the root to each node (top-down pass).
  annotated_nodes <- list()
  
  # Bottom up (compute partitions, we store the branch lengths to compute distances 
  # to the root on the way down).
  for(i in 1:num_edges) {
    
    parent <- edges[i,1]
    child <- edges[i,2]
    
    # Initialization (leaves).
    if(child <= num_leaves) {
      # We translate the index for the sorted labels.
      child <- tip_order[child]
      # Leaves have as children themselves.
      annotated_nodes[[child]] <- list(root_distance=NULL, edges_to_root=1, partitions=list(child)) 
    }
    
    # Aggregate the children partitions (only if we are not visiting a leaf).
    aggregated_partitions <- annotated_nodes[[child]]$partitions[[1]]
    if((child > num_leaves)) {
      for(p in 2:length(annotated_nodes[[child]]$partitions))
        aggregated_partitions <- c(aggregated_partitions, annotated_nodes[[child]]$partitions[[p]])
    }
    
    # Update the branch length on the child.
    annotated_nodes[[child]]$root_distance <- edge_lengths[i]
    
    # We have not visited this internal node before.
    if(parent > length(annotated_nodes) || is.null(annotated_nodes[[parent]])) {
      # Assume the first time we get the left child partition.
      annotated_nodes[[parent]] <- list(root_distance=NULL, edges_to_root=1, partitions=list(aggregated_partitions))
    }
    # This is not the first time we have visited the node.
    else {
      # We store the next partition of leaves.
      annotated_nodes[[parent]]$partitions[[length(annotated_nodes[[parent]]$partitions)+1]] <- aggregated_partitions
    }
  }
  
  # Update the distance to the root at the root (i.e. 0)
  # And the number of edges to the root (i.e. 0).
  annotated_nodes[[num_leaves+1]]$root_distance <- 0
  annotated_nodes[[num_leaves+1]]$edges_to_root <- 0
  
  # Top down, compute distances to the root for each node.
  for(i in num_edges:1) {
    parent <- edges[i,1]
    child <- edges[i,2]
    
    # If the child is a leaf we translate the index for the sorted labels.
    if(child <= num_leaves)
      child <- tip_order[child]
    
    annotated_nodes[[child]]$root_distance <- annotated_nodes[[child]]$root_distance + annotated_nodes[[parent]]$root_distance
    annotated_nodes[[child]]$edges_to_root <- annotated_nodes[[child]]$edges_to_root + annotated_nodes[[parent]]$edges_to_root
  }
  
  # Distance vectors
  vector_length <- (num_leaves*(num_leaves-1)/2) + num_leaves
  length_root_distances <- double(vector_length)
  topological_root_distances <- integer(vector_length)
  
  # Fill-in the leaves (notice the involved index translation for leaves).
  topological_root_distances[(vector_length-num_leaves+1):vector_length] <- 1
  length_root_distances[(vector_length-num_leaves+1):vector_length] <- edge_lengths[match(1:num_leaves, edges[,2])][order(tree$tip.label)]
  
  # Instead of computing the lexicographic order for the combination pairs assume we
  # are filling in a symmetric distance matrix (using only the triangular upper part).
  # We just need to "roll" the matrix indices into the vector indices.
  # Examples for (k=5)
  # The combination c(1,4) would be located at position 3 on the vector.
  # The combination c(2,1) would be located at position 1 on the vector because d(2,1) = d(1,2).
  # The combination c(2,3) would be located at position 5 on the vector.
  
  index_offsets <- c(0, cumsum((num_leaves-1):1))
  
  # This is the slow part, we compute both vectors as gain would be marginal.
  sapply(annotated_nodes, function(node) {
    
    # We skip leaves and the root (if the MRCA for M groups of leaves is at the root
    # all combinations of leaves -among different groups- have 0 as distance to the root).
    # For large trees this can spare us of computing a lot of combinations. 
    # Example: In a perfectly balanced binary tree (N/2 leaves at each side of the root), 
    # at the root we'd save (N/2) * (N/2) combinations to update. Worst case scenario is completely 
    # unbalanced tree (N-1,1), we'd save in that case only N-1 combinations.
    
    # Make sure we are not visiting a leaf or the root.
    if(length(node$partitions) > 1 && node$root_distance > 0) {
      
      # Update all combinations for pairs of leaves from different groups.
      num_groups <- length(node$partitions)
      for(group_a in 1:(num_groups-1)) {
        for(group_b in (group_a+1):num_groups) {
          CPP_update_combinations(length_root_distances, topological_root_distances, node$partitions[[group_a]], 
                                  node$partitions[[group_b]], index_offsets, node$root_distance, node$edges_to_root)
        }
      }
      
    }
  })
  
  if(!return_lambda_function)
    return(lambda * length_root_distances + (1-lambda) * topological_root_distances)
  else {
    return(function(l) { 
      if(l<0 || l>1) stop("Pick lambda in [0,1]")
      return(l * length_root_distances + (1-l) * topological_root_distances) })
  }
}

require(Rcpp)

CPP_update_combinations <- cppFunction("void updateDistancesWithCombinations(NumericVector& length_root_distances,
                                       NumericVector& topological_root_distances,
                                       IntegerVector& left_partition,
                                       IntegerVector& right_partition,
                                       IntegerVector& index_offsets,
                                       double distance_to_root,
                                       int edges_to_root) 
                                       {
                                       // Iterate through all combinations.
                                       for(int i=0; i < left_partition.size(); ++i) {
                                       for(int j=0; j < right_partition.size(); ++j) {
                                       
                                       int first_leaf = left_partition[i];
                                       int second_leaf = right_partition[j];
                                       
                                       // Because of the symmetric distances.
                                       if(left_partition[i] > right_partition[j]) {
                                       first_leaf = right_partition[j];
                                       second_leaf = left_partition[i];
                                       }
                                       
                                       // Roll the index (notice we take into account C++ indices here, starting at 0).
                                       int combination_index = index_offsets[first_leaf-1] + (second_leaf - first_leaf) - 1;
                                       
                                       // Update the vectors.
                                       length_root_distances[combination_index] = distance_to_root;
                                       topological_root_distances[combination_index] = edges_to_root;
                                       }
                                       }
                                       }")

require(ape)
tree <- rtree(6)
tree.vec(tree)

**Comment by thibautjombart** [comment]
Cool, works for me too. Odd.

**Comment by thibautjombart** [comment]
After merging master into newfeatures, I get:
* checking examples ... ERROR
Running examples in ‘treescape-Ex.R’ failed
The error most likely occurred in:

> base::assign(".ptime", proc.time(), pos = "CheckExEnv")
> ### Name: tree.vec
> ### Title: Tree vector function
> ### Aliases: tree.vec
> 
> ### ** Examples
> 
> ## generate a random tree
> tree <- rtree(6)
> ## topological vector of mrca distances from root:
> tree.vec(tree)
Error in .Primitive(".Call")(<pointer: (nil)>, length_root_distances,  : 
  NULL value passed as symbol address
Calls: tree.vec ... lapply -> FUN -> CPP_update_combinations -> <Anonymous>
Execution halted
* checking PDF version of manual ... OK
* DONE

Status: 1 ERROR, 1 WARNING
See
  ‘/tmp/Rtmp3ei4Dx/treescape.Rcheck/00check.log’
for details.

Error: Command failed (1)

**Comment by thibautjombart** [comment]
(I just pushed newfeatures)

**Comment by MichelleKendall** [comment]
This is now resolved on my fork, and it's passing the check. Most of the functions are now updated to their faster versions, I'm just doing them one by one to be sure. I'll let you know when it's good to push across - should be soon!

**Comment by thibautjombart** [comment]
Cool. Send a pull request as soon as ready :)

**Comment by MichelleKendall** [comment]
I have done the pull request, just looking into how to resolve conflicts now.

**Comment by thibautjombart** [comment]
I have merged and resolved conflicts. However, as of fb29b3469644b9faf33c0dfca59e05114b24140e
we get:
```r
* DONE (treescape)
Error: Failure in roxygen block beginning metrics.R:305
Mismatched braces: "@example ## generate 10 random trees, each with 6 tips
trees <- rmtree(10,6)

## pairwise distance matrix when lambda=0
multi.dist(trees)

## pairwise distance matrix as a function of lambda:
m <- multi.dist(trees, return_lambda_function=TRUE)

## evaluate at lambda=0. Equivalent to multi.dist(trees).
m0 <- m(0)

## A method to visualise these distances with MDS:
require(ade4)

## find an optimum projection of the points in 2 dimensions:
mMDS <- dudi.pco(as.dist(m0), scannf=FALSE,nf=2)
```

**Comment by thibautjombart** [comment]
Fixed as of 2e4a3ef725333d705b513e728f244bf375bae027

**Comment by MichelleKendall** [comment]
Fantastic! Thanks
